### PR TITLE
level system for voice chats

### PR DESCRIPTION
**Pull request by valivia**

### CODE REVIEW

**Comment by valivia** [comment]
I'd love some feedback on the xp amount and exclusion rules, maybe ignore the vc if everyone is muted? etc

File: bot/src/modules/level/lib/level.ts [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+    let guilds = state.client.guilds.cache;
+    guilds = guilds.filter((x) => state.guilds.get(x.id)?.levelVCEnabled);
+
+    for (const guild of guilds.values()) {
+      const channels = guild.channels.cache.filter((x) => x.type === ChannelType.GuildVoice) as Collection<string, VoiceChannel>;
+
+      for (const channel of channels.values()) {
+        const members = channel.members;
+        // Filter out bots and deafened users.
+        members.filter((x) => !x.user.bot);
```
Comment (luksab) [inline bot/src/modules/level/lib/level.ts:138]: Why have two different styles of filter here? (One with "(x)", and one with "member"?)

File: bot/src/modules/level/lib/level.ts [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+        members.filter(member => member.voice.deaf);
```
Comment (luksab) [inline bot/src/modules/level/lib/level.ts:139]: This will only leave deaf members, not remove them.

File: bot/src/modules/level/lib/level.ts [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+        // filter out users that are alone.
+        if (members.size <= 1) continue;
+
+        for (const member of members.values()) {
+          // TODO, remove this, temporary debug.
+          localState.log.debug(`Vcloop triggered, User ${member.user.tag} in ${guild.name}.`);
```
Comment (luksab) [inline bot/src/modules/level/lib/level.ts:146]: Probably should remove before merging.

File: bot/src/modules/level/lib/level.ts [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+    if (!this.isUserTimedOut(member, config.timeout)) return;
+
+    const xp = config.xp || LevelController.getRandomXP(guildConfig.levelModifier);
+    const data = await this.ChangeUserXP(member.guild.id, member.id, xp);
+
+    this.setUserTimeout(member);
+
+    // TODO, remove this, temporary debug.
+    if (!message) localState.log.debug(`User ${member.user.tag} in ${member.guild.name} gained ${xp} XP.`);
```
Comment (luksab) [inline bot/src/modules/level/lib/level.ts:169]: Remove as well.

**Review by luksab — CHANGES_REQUESTED** [review]
Giving a set XP every 5 minutes seems like a good way to handle voice XP. The 20 value is ofc subject to possible tuning.
It should be possible to have more and more complicated local state about when people joined/left/deafened in a VC and not have a loop at all. I think this solution is probably still better, even if more resource intensive, because it's likely much simpler.

**Review by luksab — APPROVED** [review]
LGTM 🚀

File: bot/src/modules/level/lib/level.ts [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        // Filter out bots and deafened users.
+        const members = channel.members
+          .filter(member => !member.user.bot)
+          .filter(member => !member.voice.deaf);
+
+        // filter out users that are alone.
+        if (members.size <= 1) continue;
+
+        for (const member of members.values()) {
+          await this.handleXPEvent(member, null, { xp: 20, timeout: 5 * 60 * 1000 });
```
Comment (fatalerrorcoded) [inline bot/src/modules/level/lib/level.ts:162]: I assume treating voice XP the same as message XP when it comes to timeouts is intentional?

File: bot/src/modules/level/commands/level_config/toggle.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+    state.guilds.set(guild.id, guild);
+
+    const embed = embedTemplate();
+    embed.setTitle("Level system configuration");
+    embed.setDescription(
+      `Level system enabled: ${guild.levelSystemEnabled ? "✅" : "❌"}\n` +
+      `Message experience gain: ${guild.levelMessageXPGain ? "✅" : "❌"}${!guild.levelSystemEnabled ? "*" : ""}\n` +
+      `Voice experience gain: ${guild.levelVoiceXPGain ? "✅" : "❌"}${!guild.levelSystemEnabled ? "*" : ""}` +
+      (!guild.levelSystemEnabled ? "\n*\\*Level system must be enabled for this to take effect.*" : "")
```
Comment (fatalerrorcoded) [inline bot/src/modules/level/commands/level_config/toggle.ts:59]: Maybe hide this text when neither of the gain options are enabled?

File: bot/src/modules/level/lib/level.ts (hunk 5 above)
Comment (valivia) [reply bot/src/modules/level/lib/level.ts:162]: yeah I figured this would prevent people from getting double XP and simplifies the state

**Pull request merged**